src/live_restore/live_restore_state.c

```diff
@@ -0,0 +1,362 @@
+ */
+static void
+__live_restore_state_to_string(WTI_LIVE_RESTORE_STATE state, char *state_strp)
+{
+    switch (state) {
+    case WTI_LIVE_RESTORE_STATE_NONE:
+        strcpy(state_strp, "WTI_LIVE_RESTORE_STATE_NONE");
+        break;
+    case WTI_LIVE_RESTORE_STATE_BACKGROUND_MIGRATION:
+        strcpy(state_strp, "WTI_LIVE_RESTORE_STATE_BACKGROUND_MIGRATION");
+        break;
+    case WTI_LIVE_RESTORE_STATE_CLEAN_UP:
+        strcpy(state_strp, "WTI_LIVE_RESTORE_STATE_CLEAN_UP");
+        break;
+    case WTI_LIVE_RESTORE_STATE_COMPLETE:
+        strcpy(state_strp, "WTI_LIVE_RESTORE_STATE_COMPLETE");
+        break;
+    }
+}
+
+/*
+ * __live_restore_state_from_string --
+ *     Convert a string to its live restore state.
+ */
+static int
+__live_restore_state_from_string(
+  WT_SESSION_IMPL *session, char *state_str, WTI_LIVE_RESTORE_STATE *statep)
+{
+    if (strcmp(state_str, "WTI_LIVE_RESTORE_STATE_NONE") == 0)
+        *statep = WTI_LIVE_RESTORE_STATE_NONE;
+    else if (strcmp(state_str, "WTI_LIVE_RESTORE_STATE_BACKGROUND_MIGRATION") == 0)
+        *statep = WTI_LIVE_RESTORE_STATE_BACKGROUND_MIGRATION;
+    else if (strcmp(state_str, "WTI_LIVE_RESTORE_STATE_CLEAN_UP") == 0)
+        *statep = WTI_LIVE_RESTORE_STATE_CLEAN_UP;
+    else if (strcmp(state_str, "WTI_LIVE_RESTORE_STATE_COMPLETE") == 0)
+        *statep = WTI_LIVE_RESTORE_STATE_COMPLETE;
+    else
+        WT_RET_MSG(session, EINVAL, "Invalid state string: '%s' ", state_str);
+
+    return (0);
+}
+
+/*
+ * __live_restore_get_state_from_file --
+ *     Read the live restore state from the turtle file. If it doesn't exist we return NONE. The
+ *     caller must already hold the live restore state lock. This function takes a non-live restore
+ *     file system as it can be called in non-live restore modes.
+ */
+static int
+__live_restore_get_state_from_file(
+  WT_SESSION_IMPL *session, WT_FILE_SYSTEM *fs, WTI_LIVE_RESTORE_STATE *statep)
+{
+    WT_DECL_RET;
+
+    bool turtle_in_dest = false;
+    WTI_LIVE_RESTORE_STATE state_from_file = WTI_LIVE_RESTORE_STATE_NONE;
+    char *lr_metadata = NULL;
+
+    WT_DECL_ITEM(dest_turt_path);
+    WT_RET(__wt_scr_alloc(session, 0, &dest_turt_path));
+    WT_ERR(__wt_filename_construct(
+      session, S2C(session)->home, WT_METADATA_TURTLE, UINTMAX_MAX, UINT32_MAX, dest_turt_path));
+
+    WT_ERR(
+      fs->fs_exist(fs, (WT_SESSION *)session, (char *)(dest_turt_path)->data, &turtle_in_dest));
+
+    if (!turtle_in_dest)
+        state_from_file = WTI_LIVE_RESTORE_STATE_NONE;
+    else {
+        WT_ERR_NOTFOUND_OK(
+          __wt_metadata_search(session, WT_METADATA_LIVE_RESTORE, &lr_metadata), true);
+        if (ret == WT_NOTFOUND) {
+            state_from_file = WTI_LIVE_RESTORE_STATE_NONE;
+            ret = 0;
+        } else {
+            char lr_metadata_string[WT_LIVE_RESTORE_STATE_STRING_MAX];
+            if ((sscanf(lr_metadata, "state=%127s", lr_metadata_string)) != 1)
```

Comment:
Instead of hardcoding 127, what about using `%.*s` and then providing the length from the #define macro you already have.

Reply:
I had a look into this one and `%.*s` only seems to be supported for `printf`, not `scanf`. 
We could build the string before calling scanf here if that's not too much overkill

Reply:
Maybe add an assert before the sscanf about the value of the `WT_LIVE_RESTORE_STATE_STRING_MAX` so that it will trigger if someone changes that constant.

Reply:
ecf736108